Nonprofit-Exchange-Hub/web-app
# Backend user routes

Apologies for the mess of a PR - at some point the `/dist` directory snuck into my git commits so deleting them takes up a ton of the files changed.

I rebuilt the user routes to get a feel for how NestJS works. They are now connected to our postgres db, and you can perform all CRUD operations on /api/users. I did not alter existing Auth routes, but will be doing so next to implement bcrypt encryption. Made minor change to test files to prevent failure, but they will eventually need a mock of TypeORM's import in the service files.

To test: 

- Create a user in db (doesn't require authentication) 
 ```
curl --request POST \
  --url http://localhost:3001/api/users \
  --header 'Content-Type: application/json' \
  --data '{
	"first_name": "TestFirst",
	"last_name": "TestLast",
	"email": "test@test.com",
	"password": "testpassword"
	}'
```
returns `{"first_name":"TestFirst","last_name":"TestLast","email":"test@test.com","password":"testpassword","id":1}`

- Get users without auth token should fail 
```
curl --request GET \
  --url http://localhost:3001/api/users 
```
returns `{"statusCode":401,"message":"Unauthorized"}`

- Login to get access token
```
curl --request POST \
  --url http://localhost:3001/api/auth/login \
  --header 'Content-Type: application/json' \
  --data '{
	"username": "TestFirst",
	"password": "TestPass"
}'
```
returns an an object with an access_token property like `{"access_token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE2MjU5NDg1MTUsImV4cCI6MTYyNTk0ODU3NX0.VZb9X5Wkz0mxlN2hK32dcbKgP2h-BbQMZ39HPA0l2jA"}`

- Get users using the access_token should return all users. Replace the access token with the one you generated by /auth/login.
```
curl --request GET \
  --url http://localhost:3001/api/users \
  --header 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE2MjU5NDg2NTQsImV4cCI6MTYyNTk0ODcxNH0.xakTSjoaOay8eRLFhgsu9LztsXakeqamrPaCbB-rWYg'
```
returns all users. In my case `[{"id":1,"first_name":"TestFirst","last_name":"TestLast","email":"test@test.com","password":"testpassword"},{"id":2,"first_name":"TestFirst","last_name":"TestLast","email":"test@test.com","password":"testpassword"},{"id":3,"first_name":"TestFirst","last_name":"TestLast","email":"test@test.com","password":"testpassword"},{"id":4,"first_name":"TestFirst","last_name":"TestLast","email":"test@test.com","password":"testpassword"}]`

Should be able to perform GET user by an id, PATCH user, and DELETE user as long as you have a valid token.
